tempesta_fw/sched/tfw_sched_hash.c

```diff
@@ -274,26 +288,32 @@ __add_conn(TfwHashConnList *cl, TfwSrvConn *conn, unsigned long hash)
 }
 
 static void
-__fill_srv_lists(TfwHashConnList *cl)
+__fill_srv_lists(TfwHashConnList *cl, size_t srv_n, TfwHashSrvConnList *srv_cls)
 {
-	size_t i;
-
-	for (i = 0; i < cl->conn_n; ++i) {
-		TfwHashConn *hconn = &cl->conns[i];
-		TfwServer *srv = (TfwServer *)hconn->conn->peer;
-		TfwHashConnList *scl = srv->sched_data;
-
-		scl->conns[scl->conn_n] = *hconn;
-		++scl->conn_n;
+	size_t i, j;
+
+	for (i = 0; i < srv_n; ++i) {
+		TfwServer *srv = srv_cls[i].srv;
+		TfwHashConnList *scl = srv_cls[i].cl;
+
+		for (j = 0; j < cl->conn_n; ++j) {
+			TfwHashConn *hconn = &cl->conns[j];
+			if ((TfwServer *)hconn->conn->peer == srv) {
```

Comment:
`sched_data` fileld in `TfwServer` and `TfwSrvGroup` is used to store data required for scheduler. During live reconfiguration schedulers are busy, that means at any time `sched_data` must be valid.
If we need to add or remove server from group we have to stop group's scheduler. `sched_data` will be set to `NULL` for both `TfwServer` and `TfwSrvGroup`. 

This function copies parts of combined server group scheduling data to per-server arrays. It's not possible to prepare complete and valid per-server scheduling data in one pass, since `TfwServer` cannot be linked with it's scheduling data until data is valid. Thats why nested loops are used.